source/Cosmos.System2_Plugs/System/MathImpl.cs

```diff
@@ -187,55 +202,163 @@ public static double Cos(double x)
         }
-        #endregion
+
+        #endregion Cosh
 
         #region Exp
+
+        /*
+        * ====================================================
+        * Copyright (C) 2004 by Sun Microsystems, Inc. All rights reserved.
+        *
+        * Permission to use, copy, modify, and distribute this
+        * software is freely granted, provided that this notice
+        * is preserved.
+        * ====================================================
+        */
+
+        // Look at http://www.netlib.org/fdlibm/e_exp.c for more a in deth explanation
+        private static int HighWord(double x)
+        {
+            long value = BitConverter.DoubleToInt64Bits(x);
+            Byte[] valueBytes = BitConverter.GetBytes(value);
+            int offset = BitConverter.IsLittleEndian ? 4 : 0;
+            return BitConverter.ToInt32(valueBytes, offset);
+        }
+
+        private static int LowWord(double x) //Opposite of high word
+        {
+            long value = BitConverter.DoubleToInt64Bits(x);
+            Byte[] valueBytes = BitConverter.GetBytes(value);
+            return BitConverter.ToInt32(valueBytes, BitConverter.IsLittleEndian ? 0 : 4);
+        }
+
         public static double Exp(double x)
         {
-            double c;
-            int n = 1;
-            double ex = 1F;
-            double m = 1F;
-            while (x > 10.000F)
-            {
-                m *= 22026.4657948067;
-                x -= 10F;
-            }
-            while (x > 01.000F)
-            {
-                m *= E;
-                x -= 1F;
+            double y, hi = 0, lo = 0, c, t;
+            int k = 0, xsb;
+
+            const double one = 1.0,
+                o_threshold = 7.09782712893383973096e+02,
+                u_threshold = -7.45133219101941108420e+02,
+                invln2 = 1.44269504088896338700e+00,
+                twom1000 = 9.33263618503218878990e-302,
+                P1 = 1.66666666666666019037e-01,
+                P2 = -2.77777777770155933842e-03,
+                P3 = 6.61375632143793436117e-05,
+                P4 = -1.65339022054652515390e-06,
+                P5 = 4.13813679705723846039e-08,
+                huge = 1.0e+300;
+
+            int hx = HighWord(x); //Highword of x
+
+            xsb = ((int)hx >> 31) & 1; //Get sign of x
+            hx &= 0x7fffffff; //Get the abs(x) of the highword
+
+            //Check if non-finite argument
+            if (hx >= 0x40862E42)
+            {           /* if |x|>=709.78... */
+                if (hx >= 0x7ff00000)
+                {
+                    Console.WriteLine("Low word! Text");
+                    if (((hx & 0xfffff) | LowWord(x)) != 0) //Assume that __Lo(x) is lower word of x
+                        return x;       /* NaN */
+                    else return (xsb == 0) ? x : 0.0;   /* exp(+-inf)={inf,0} */
```

Comment:
Can you changes this and other occurences to have the body of the if on the next line?